### PR TITLE
WIP - start adding optional overwrite to insert operation (RepSert)

### PR DESCRIPTION
**Pull request by ObiWahn**
This PR will implement REPSERT a variation of UPSERT for the http interface.

More information on the feature can be found in the [RestUpsert proposal](https://github.com/arangodb/documents/blob/RestUpsert/FeatureProposals/RestUpsert.md)

### CODE REVIEW

File: arangod/Transaction/Methods.cpp [hunk 1]
```diff
@@ -1532,10 +1545,9 @@ OperationResult transaction::Methods::insertLocal(
       // Now replicate the good operations on all followers:
       std::string path =
         "/_db/" + arangodb::basics::StringUtils::urlEncode(databaseName()) +
-        "/_api/document/" +
-        arangodb::basics::StringUtils::urlEncode(collection->name()) +
-        "?isRestore=true&isSynchronousReplication=" +
-        ServerState::instance()->getId();
+        "/_api/document/" + arangodb::basics::StringUtils::urlEncode(collection->name()) +
+        "?isRestore=true&isSynchronousReplication=" + ServerState::instance()->getId() +
+        "&" + StaticStrings::OverWrite + "=" + (options.overwrite ? "true" : "false"); //CHEKME
```
Comment (danielhlarkin) [inline arangod/Transaction/Methods.cpp]: Would it make more sense to use the PUT API, which is meant for replacements, instead of adding an extra option via URL to the creation API? I haven't looked too in-depth at what's there. We could also handle a more natural UPSERT with property merging via the PATCH API, maybe?

File: arangod/Transaction/Methods.cpp [hunk 2]
```diff
@@ -1466,10 +1466,23 @@ OperationResult transaction::Methods::insertLocal(
     ManagedDocumentResult result;
     TRI_voc_tick_t resultMarkerTick = 0;
     TRI_voc_rid_t revisionId = 0;
+    auto needsLock = [this,&collection]() -> bool {return !isLocked(collection, AccessMode::Type::WRITE);};
+
+    Result res = collection->insert( this, value, result, options
+                                   , resultMarkerTick, needsLock(), revisionId
+                                   );
```
[the remark this answers is not in the document]
Comment (ObiWahn) [reply arangod/Transaction/Methods.cpp]: I think the lambda is called twice and I was not sure if the lock status could change in the meantime.

File: arangod/Transaction/Methods.cpp (hunk 1 above)
Comment (ObiWahn) [reply arangod/Transaction/Methods.cpp]: We have discussed this long with @neunhoef and he was sure that POST is the thing to do.

File: Documentation/DocuBlocks/Rest/Documents/post_create_document.md [hunk 3]
```diff
@@ -31,6 +31,11 @@ If set to *true*, an empty object will be returned as response. No meta-data
 will be returned for the created document. This option can be used to
 save some network traffic.
 
+@RESTQUERYPARAM{overwrite,boolean,optional}
+If set to *true*, the insert becomes a replace-insert. If a document with the
+same *_key* already existst the new document is not rejected with unique
```
Comment (jsteemann) [inline Documentation/DocuBlocks/Rest/Documents/post_create_document.md]: exists

File: Documentation/DocuBlocks/Rest/Documents/post_create_document.md [hunk 4]
```diff
@@ -239,5 +244,27 @@ Use of returnNew:
...
+    var body = '{ "Hello": "World", "_key" : "lock" }';
+    var response = logCurlRequest('POST', url, body);
+    // insert
+    assert(response.code === 201);
+    logJsonResponse(response);
+
+    url = "/_api/document/" + cn + "?overwrite=true";
+    response = logCurlRequest('POST', url, body);
+    // insert same key
+    assert(response.code === 201);
```
Comment (jsteemann) [inline Documentation/DocuBlocks/Rest/Documents/post_create_document.md]: Should an overwrite actually return HTTP 201 (Created)? or should it rather return HTTP 200 (Ok)?
How can the caller distinguish between insert and replace here?

File: CHANGELOG [hunk 5]
```diff
@@ -1,6 +1,12 @@
 devel
 -----
 
+* added `overwrite` option to the `document rest-handler` to allow for easier syncing.
```
Comment (jsteemann) [inline CHANGELOG]: to the `insert` operation of the document API REST handler

File: CHANGELOG [hunk 6]
```diff
@@ -1,6 +1,12 @@
...
+
+  This implements almost the much inquired UPSERT. In reality it is a REPSERT
+  (replace/insert) because modification of documents is not possible. The option does
+  not in cluster collections with custom sharding.
```
Comment (jsteemann) [inline CHANGELOG]: does not what?

File: CHANGELOG [hunk 7]
```diff
@@ -1,6 +1,12 @@
 devel
 -----
 
+* added `overwrite` option to the `document rest-handler` to allow for easier syncing.
+
+  This implements almost the much inquired UPSERT. In reality it is a REPSERT
+  (replace/insert) because modification of documents is not possible. The option does
```
Comment (jsteemann) [inline CHANGELOG]: modification of documents is possible!

File: arangod/Transaction/Methods.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/Transaction/Methods.cpp]: No, the lock status will not change in the meantime. It is ok to remove the lambda and store the result in a const boolean.

File: arangod/Transaction/Methods.cpp [hunk 8]
```diff
@@ -1466,10 +1466,23 @@ OperationResult transaction::Methods::insertLocal(
...
 
-    Result res =
-        collection->insert(this, value, result, options, resultMarkerTick,
-                           !isLocked(collection, AccessMode::Type::WRITE), revisionId);
+    if(options.overwrite && res.is(TRI_ERROR_ARANGO_UNIQUE_CONSTRAINT_VIOLATED)){
+      // RepSert Case - unique_constraint violated -> maxTick has not changed -> try replace
```
Comment (jsteemann) [inline arangod/Transaction/Methods.cpp]: could be a unique constraint violation on a secondary index. does this need to be differentiated?

File: arangod/Transaction/Methods.cpp [hunk 9]
```diff
@@ -1466,10 +1466,23 @@ OperationResult transaction::Methods::insertLocal(
...
+      resultMarkerTick = 0;
+      TRI_voc_rid_t revision;
+      ManagedDocumentResult previousDocumentResult; // return OLD/NEW?
+      _localHints.unset(transaction::Hints::Hint::SINGLE_OPERATION);
```
Comment (jsteemann) [inline arangod/Transaction/Methods.cpp]: the "unset" should not be necessary here, and it indicates something is wrong at the call site.
Any caller should not have set the "SINGLE_OPERATION" hint in case an overwrite operation is attempted.
We should even add an assert that the "SINGLE_OPERATION" hint is not set when an overwrite is attempted.

Comment (jsteemann) [reply arangod/Transaction/Methods.cpp]: Removing the hint just here may have bad side effects, which we would like to avoid if possible.

**Comment by jsteemann** [comment]
Please add more tests, e.g. for baby operations in the `server_http` test target, and for `http_server`.
Please add tests that verify the claimed behavior in cluster case (custom shard keys used) and for unique constraint violations on secondary indexes.
Please document the new "overwrite" parameter in the document operations section of the manual.

**Comment by jsteemann** [comment]
How is the repsert functionality used from AQL?

File: Documentation/DocuBlocks/Rest/Documents/post_create_document.md [hunk 10]
```diff
@@ -239,5 +244,27 @@ Use of returnNew:
...
+
+    var url = "/_api/document/" + cn;
+    var body = '{ "Hello": "World", "_key" : "lock" }';
+    var response = logCurlRequest('POST', url, body);
+    // insert
+    assert(response.code === 201);
+    logJsonResponse(response);
+
+    url = "/_api/document/" + cn + "?overwrite=true";
+    response = logCurlRequest('POST', url, body);
```
Comment (dothebart) [inline Documentation/DocuBlocks/Rest/Documents/post_create_document.md]: shouldn't you rather write another document with the same id here then completely the same again?

File: Documentation/DocuBlocks/Rest/Documents/post_create_document.md (hunk 4 above)
Comment (ObiWahn) [reply Documentation/DocuBlocks/Rest/Documents/post_create_document.md]: The caller will get the old revision id in the returned object. Lets keep the options that modify the return code to a minimum so the API does not become unpredictable.

File: Documentation/DocuBlocks/Rest/Documents/post_create_document.md (hunk 10 above)
Comment (ObiWahn) [reply Documentation/DocuBlocks/Rest/Documents/post_create_document.md]: The document has the "_key" already defined. Therefor posting the same document twice triggers the behavior that is meant to be shown.

File: Documentation/Books/Manual/DataModeling/Documents/DocumentMethods.md [hunk 11]
```diff
@@ -1,5 +1,6 @@
 Collection Methods
 ==================
+<!-- the filename is a fail - can it be moved? -->
```
Comment (jsteemann) [inline Documentation/Books/Manual/DataModeling/Documents/DocumentMethods.md]: There are methods to work with documents on the database level and on the collection level.
"Collection Methods" here means "these are the functions to work with documents via the collection object".

Can we remove this comment please?

File: Documentation/Books/Manual/DataModeling/Documents/DocumentMethods.md [hunk 12]
```diff
@@ -567,8 +568,15 @@ used to specify the following options:
     a default *waitForSync* value of *true*.
   - *silent*: If this flag is set to *true*, the method does not return
     any output.
+  - *overwirte*: If set to *true*, the insert becomes a replace-insert.
```
Comment (jsteemann) [inline Documentation/Books/Manual/DataModeling/Documents/DocumentMethods.md]: the option is named "overwrite", not "overwirte".

File: Documentation/DocuBlocks/Rest/Documents/post_create_document.md [hunk 13]
```diff
@@ -26,11 +26,20 @@ Wait until document has been synced to disk.
 Additionally return the complete new document under the attribute *new*
 in the result.
 
+@RESTQUERYPARAM{returnOld,boolean,optional}
+Additionally return the complete old document under the attribute *old*
+in the result. Only available if the overwirte option is used.
```
Comment (jsteemann) [inline Documentation/DocuBlocks/Rest/Documents/post_create_document.md]: overwrite!

File: UnitTests/HttpInterface/api-document-create-spec.rb [hunk 14]
```diff
@@ -845,8 +845,204 @@
...
+        key.should be_kind_of(String)
+
+        match = didRegex.match(did)
+
+        match[1].should eq("#{@cn}")
+
+        etag.should eq("\"#{rev}\"")
+        location.should eq("/_db/_system/_api/document/#{did}")
+
+        cmd = "/_api/document?collection=#{@cn}&overwrite=true&returnNew=true&returnOld=true&waitForSync=true"
```
Comment (jsteemann) [inline UnitTests/HttpInterface/api-document-create-spec.rb]: "returnOld" is set here, but the test does not check whether the old document is part of the result.

File: arangod/Transaction/Helpers.cpp [hunk 15]
```diff
@@ -385,18 +385,18 @@ OperationResult transaction::helpers::buildCountResult(std::vector<std::pair<std
     }
     resultBuilder.close();
   }
-  return OperationResult(Result(), resultBuilder.steal(), nullptr, false);
+  return OperationResult(Result(), resultBuilder.buffer(), nullptr);
```
Comment (jsteemann) [inline arangod/Transaction/Helpers.cpp]: is this change intentional?

File: arangod/Transaction/Methods.cpp [hunk 16]
```diff
@@ -1557,12 +1588,10 @@ OperationResult transaction::Methods::insertLocal(
...
-        "/_db/" +
-        arangodb::basics::StringUtils::urlEncode(vocbase().name()) +
-        "/_api/document/" +
-        arangodb::basics::StringUtils::urlEncode(collection->name()) +
-        "?isRestore=true&isSynchronousReplication=" +
-        ServerState::instance()->getId();
+        "/_db/" + arangodb::basics::StringUtils::urlEncode(vocbase().name()) +
+        "/_api/document/" + arangodb::basics::StringUtils::urlEncode(collection->name()) +
+        "?isRestore=true&isSynchronousReplication=" + ServerState::instance()->getId();
+        "&" + StaticStrings::OverWrite + "=" + (options.overwrite ? "true" : "false"); //CHEKME
```
Comment (jsteemann) [inline arangod/Transaction/Methods.cpp]: why "CHEKME" here?

File: arangod/Transaction/Methods.h [hunk 17]
```diff
@@ -414,6 +414,8 @@ class Methods {
   /// @brief build a VPack object with _id, _key and _rev and possibly
   /// oldRef (if given), the result is added to the builder in the
   /// argument as a single object.
+
+  //SHOULD THE OPTIONS BE CONST?
```
Comment (jsteemann) [inline arangod/Transaction/Methods.h]: which options?

File: arangod/Utils/OperationResult.h [hunk 18]
```diff
@@ -26,23 +26,27 @@
...
+
 struct OperationResult {
-  OperationResult() {} 
-  
+  OperationResult() {}
+
   // create from integer status code
   explicit OperationResult(int code) : result(code) {}
-  
+  explicit OperationResult(int code, OperationOptions const& options) : result(code), _options(options) {};
```
Comment (jsteemann) [inline arangod/Utils/OperationResult.h]: semicolon at end of line is redundant

File: js/client/tests/shell/shell-document.js [hunk 19]
```diff
@@ -30,6 +30,7 @@
 
 var jsunity = require("jsunity");
 var arangodb = require("@arangodb");
+var ERRORS = arangodb.errors;
```
Comment (jsteemann) [inline js/client/tests/shell/shell-document.js]: what is this used for?

File: js/common/tests/shell/shell-document.js [hunk 20]
```diff
@@ -2338,6 +2338,44 @@ function DatabaseDocumentSuiteReturnStuff () {
...
+      try{
+        var res = collection.insert({a : 2, _key : key});
+        fail();
+      }
+      catch (err) {
+        assertEqual(ERRORS.ERROR_ARANGO_UNIQUE_CONSTRAINT_VIOLATED.code, err.errorNum);
+      }
+
+      // overwrite with same key must work
+      collection.insert({a : 2, _key: key},{overwrite:true});
```
Comment (jsteemann) [inline js/common/tests/shell/shell-document.js]: the value of "a" should actually be changed here, in order to see that the document was actually replaced!
Apart from that, we want to verify that the document got actually replaced, so it would be good to write *new* attributes here, and test that some *old* attributes actually vanished after carrying out the replace.

File: js/common/tests/shell/shell-document.js [hunk 21]
```diff
@@ -2338,6 +2338,44 @@ function DatabaseDocumentSuiteReturnStuff () {
...
+      assertEqual(arr[0].a, 2);
+
+      // overwrite (babies) with same key must work
+      collection.insert({a : 2, _key: key},{overwrite:true});
+      arr = collection.toArray();
+      assertEqual(arr.length, 1);
+      assertEqual(arr[0].a, 2);
+
+      // overwrite (babies) with same key must work
+      collection.insert([{a : 3, _key: key}, {a : 4, _key: key}, {a : 5, _key: key}], {overwrite:true});
```
Comment (jsteemann) [inline js/common/tests/shell/shell-document.js]: same here: we should test that it's actually replace, not an update!

Comment (jsteemann) [reply js/common/tests/shell/shell-document.js]: additionally a test is missing for a repsert (overwrite: true) for which the document does not initially exist.

Comment (jsteemann) [reply js/common/tests/shell/shell-document.js]: should we add a test for a repsert (overwrite: true) when the inserted document does not contain a `_key` attirubte?

**Comment by jsteemann** [comment]
Test in server_http are still missing.

**Comment by jsteemann** [comment]
Additionally there is no AQL integration for the feature, and tests for AQL queries with the "overwrite" flag.

File: Documentation/Books/Manual/DataModeling/Documents/DocumentMethods.md (hunk 11 above)
Comment (ObiWahn) [reply Documentation/Books/Manual/DataModeling/Documents/DocumentMethods.md]: There is no section "Document Methods" within the file. All calls are methods of collections even though they modify / create / delete documents. I think it would be better to rename the file so it matches it's content. But OK.

File: arangod/Transaction/Helpers.cpp (hunk 15 above)
Comment (ObiWahn) [reply arangod/Transaction/Helpers.cpp]: yes

File: arangod/Transaction/Methods.h (hunk 17 above)
Comment (ObiWahn) [reply arangod/Transaction/Methods.h]: The OperationOptions that get passed into the following calls.

File: UnitTests/HttpInterface/api-document-create-spec.rb (hunk 14 above)
Comment (ObiWahn) [reply UnitTests/HttpInterface/api-document-create-spec.rb]: in 967 the old document is inspected

File: UnitTests/HttpInterface/api-document-create-spec.rb [hunk 22]
```diff
@@ -845,8 +845,204 @@
...
+        newrev.should !eq(rev)
+
+        newoldrev = newdoc.parsed_response['_oldRev']
+        newoldrev.should be_kind_of(String)
+        newoldrev.should eq(rev)
+
+        newkey = newdoc.parsed_response['_key']
+        newkey.should be_kind_of(String)
+        newkey.should eq(key)
+        newdoc.code.should eq(202)
```
Comment (jsteemann) [inline UnitTests/HttpInterface/api-document-create-spec.rb]: btw we could check for the updated value of "Hallo" here. It would make sense.

File: UnitTests/HttpInterface/api-document-create-spec.rb [hunk 23]
```diff
@@ -845,8 +845,204 @@
...
+
+        newnew = newdoc.parsed_response['new']
+        newnew["_key"].should be_kind_of(String)
+        newnew["_key"].should eq(key)
+        newnew["_rev"].should eq(newrev)
+
+        newold = newdoc.parsed_response['old']
+        newold["_key"].should eq(key)
+        newold["_rev"].should eq(newoldrev)
+        newdoc.code.should eq(201)
```
Comment (jsteemann) [inline UnitTests/HttpInterface/api-document-create-spec.rb]: we should also check for the updated value of "Hallo" here.

File: UnitTests/HttpInterface/api-document-create-spec.rb [hunk 24]
```diff
@@ -845,8 +845,204 @@
...
+        newkey.should eq(key)
+
+        newnew = newdoc.parsed_response[0]['new']
+        newnew["_key"].should be_kind_of(String)
+        newnew["_key"].should eq(key)
+        newnew["_rev"].should eq(newrev)
+
+        newold = newdoc.parsed_response[0]['old']
+        newold["_key"].should eq(key)
+        newold["_rev"].should eq(newoldrev)
```
Comment (jsteemann) [inline UnitTests/HttpInterface/api-document-create-spec.rb]: we should check for the value of "Hallo" here too.

File: arangod/Transaction/Helpers.cpp (hunk 15 above)
Comment (jsteemann) [reply arangod/Transaction/Helpers.cpp]: why is it needed?

File: arangod/Transaction/Methods.h (hunk 17 above)
Comment (jsteemann) [reply arangod/Transaction/Methods.h]: They are already const... So where's is the problem?

File: js/common/tests/http/document-babies-spec.js [hunk 25]
```diff
@@ -1208,4 +1208,77 @@ describe('babies collection document', function () {
...
+      expect(b2).to.have.lengthOf(3);
+      expect(b2[0]).to.be.a('object');
+      expect(b2[1]).to.be.a('object');
+      expect(b2[2]).to.be.a('object');
+      expect(b2[0]._key).to.be.a('string');
+      expect(b2[0]._key).to.equal(key1);
+      expect(b2[1]._key).to.equal(key1);
+      expect(b2[2]._key).to.equal(key1);
+      expect(b2[1]._rev).to.equal(b2[2].old._rev);
+      expect(b2[2].old._rev).to.equal(b2[1].new._rev);
```
Comment (jsteemann) [inline js/common/tests/http/document-babies-spec.js]: "old" is not check here.

File: js/common/tests/http/document-babies-spec.js [hunk 26]
```diff
@@ -1208,4 +1208,77 @@ describe('babies collection document', function () {
...
+        body: JSON.stringify([{
+          '_key': res1._key,
+          'ulf': 42
+        }])
+      }));
+      let b2 = JSON.parse(req2.rawBody);
+      let res2 = b2[0];
+
+      expect(req2.statusCode).to.equal(202);
+      expect(req2._key).to.equal(req1._key);
```
Comment (jsteemann) [inline js/common/tests/http/document-babies-spec.js]: "old" is not checked here.
It is not checked if the document later has the correct values in it.

File: js/common/tests/shell/shell-document.js [hunk 27]
```diff
@@ -2338,6 +2338,52 @@ function DatabaseDocumentSuiteReturnStuff () {
...
+      catch (err) {
+        assertEqual(ERRORS.ERROR_ARANGO_UNIQUE_CONSTRAINT_VIOLATED.code, err.errorNum);
+      }
+
+      // overwrite with same key must work
+      var rv = collection.insert({c : 3, _key: key},{overwrite:true, returnOld:true, returnNew:true});
+      var arr = collection.toArray();
+      assertEqual(arr.length, 1);
+      assertEqual(rv.new.c, 3);
+      assertEqual(rv.new.a, undefined);
```
Comment (jsteemann) [inline js/common/tests/shell/shell-document.js]: assertFalse(rv.new.hasOwnProperty('a'));

File: js/server/tests/aql/aql-modify-cluster.js [hunk 28]
```diff
@@ -1008,9 +1008,51 @@ function ahuacatlInsertSuite () {
...
+/// @brief test insert
+////////////////////////////////////////////////////////////////////////////////
+
+    testInsertOverwrite : function () {
+      c1.truncate();
+      assertEqual(0, c1.count());
+
+      var rv1 = db._query(" INSERT { _key: '123', name: 'ulf' } IN @@cn OPTIONS { overwrite: false } RETURN NEW", { "@cn": cn1 });
+      var resultArray1 = rv1.toArray();
+      assertEqual(1, c1.count());
```
Comment (jsteemann) [inline js/server/tests/aql/aql-modify-cluster.js]: result contents not checked here.

File: js/server/tests/aql/aql-modify-cluster.js [hunk 29]
```diff
@@ -1008,9 +1008,51 @@ function ahuacatlInsertSuite () {
...
+
+      var rv2 = db._query(" INSERT { _key: '123', name: 'ulfine' } IN @@cn OPTIONS { overwrite: true } RETURN {old: OLD, new: NEW}", { "@cn": cn1 });
+      var resultArray2 = rv2.toArray();
+
+      assertEqual(1, c1.count());
+      assertEqual(resultArray1[0]._rev, resultArray2[0].old._rev);
```
Comment (jsteemann) [inline js/server/tests/aql/aql-modify-cluster.js]: results not checked here

File: js/server/tests/aql/aql-modify-cluster.js [hunk 30]
```diff
@@ -1008,9 +1008,51 @@ function ahuacatlInsertSuite () {
...
+        `, { "@cn": cn1 });
+
+      var resultArray3 = rv3.toArray();
+
+      assertEqual(3, c1.count());
+      assertEqual(resultArray2[0].new._rev, resultArray3[0].old._rev);
+      assertEqual(resultArray2[0].new._key, resultArray3[0].old._key);
+      assertEqual(resultArray2[0].new.name, "ulfine");
+      assertEqual(resultArray3[0].old.name, "ulfine");
+      assertEqual(resultArray3[0].new.name, 3);
```
Comment (jsteemann) [inline js/server/tests/aql/aql-modify-cluster.js]: resultArray3[1] and resultArray[2] are not checked here.

**Comment by jsteemann** [comment]
Additionally, the constructor of an InsertNode from a VelocyPack Slice does not seem to honor the "old" variable at the moment and probably needs adjustment.

File: arangod/Transaction/Methods.h (hunk 17 above)
Comment (ObiWahn) [reply arangod/Transaction/Methods.h]: https://github.com/arangodb/arangodb/blob/7d5d8300702c07345d9f7f68a788495196ee35f1/arangod/Transaction/Methods.h#L428

**Comment by ObiWahn** [comment]
https://github.com/arangodb/arangodb/blob/devel/arangod/Aql/ModificationNodes.cpp#L44 "old" variable is already created here.

File: arangod/Transaction/Helpers.cpp (hunk 15 above)
Comment (ObiWahn) [reply arangod/Transaction/Helpers.cpp]: This is not necessary because OperationResult takes a `const ref` that prologs the lifetime of the temporary stolen object. The `buffer()` on the other hand does not return the object itself but a reference making it more obvious what is going on and will even work if the constructor's argument should loose it's constness.

Comment (ObiWahn) [reply arangod/Transaction/Helpers.cpp]: Further more `steal()` copies the shared_ptr for nothing.

**Comment by jsteemann** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr-linux/552/
